[Input][joy] Simplify focus with `:focus-within` and add examples (#37385)
mui/material-ui: docs/data/joy/components/input/input.md

{{"demo": "InputFormProps.js"}}

### Focused ring

Provide these CSS variables to `sx` prop to control the focused ring appearance:

- `--Input-focusedInset`: the focused ring's **position**, either inside(`inset`) or outside(`var(--any, )`) of the Input.
- `--Input-focusedThickness`: the **size** of the focused ring.
- `--Input-focusedHighlight`: the **color** of the focused ring.

{{"demo": "FocusedRingInput.js"}}

:::success
To get full control of the focused ring, customize the `box-shadow` of the `::before` pseudo element directly

```js

> <Input sx={{ '&:focus-within(::before)': { boxShadow: '...your custom value' } }} />

Reviewer: Should it be:
```suggestion
<Input sx={{ '&:focus-within::before': { boxShadow: '...your custom value' } }} />
```

Author: Nice catch!

```

:::

#### Debugging the focus ring

To display the Input's focus ring by simulating user's focus, inspect the Input element and toggle the [pseudostate panel](https://developer.chrome.com/docs/devtools/css/#pseudostates).

- If you inspect the Input's root element, with `.MuiInput-root` class, you have to toggle on the `:focus-within` state.

> - If you inspect the `<input>` element, you have to toggle on the `:focus` state.

Reviewer: We can use `:focus-within` as well for `<input>`, right?
```suggestion
- If you inspect the `<input>` element, you can toggle with either `:focus` or `:focus-within` states.
```

Author: You are right, it works!

### Triggering the focus ring

To trigger the focus ring programmatically, set the CSS variable `--Input-focused: 1`.

{{"demo": "TriggerFocusInput.js"}}

:::info

> 💡 The focus ring still appear on focus even though you set `--Input-focused: 0`.

Reviewer: Regarding this statement, I checked that if we set `--Input-focused: 0`, the focused is removed. Is this correct? The thickness is multiplied by zero.

Author: Yes, but you will see the focus ring if you focus on the input because `:focus-within` has higher CSS specificity.

https://github.com/mui/material-ui/assets/18292247/2886beb9-437a-46f1-b7ed-27c4b337a317
